### PR TITLE
Game data loading/saving optimizations.

### LINKED ISSUE #520: An error occurred while trying to load: Array dimensions exceeded supported range
Was messing around with trying to extract a non-deltarune/undertale game, and ran into this issue. Assuming the file was too large: file size 2,444,982 kb.

### PR DESCRIPTION
**Pull request by VladiStep**
## Description
**This PR is deprecated in favor of the separate PRs: #960, #962, #963, #971.**

Game data loading and saving improvements:

1. **Game data is now loaded and saved directly, without buffering whole file data in memory** - new `FileBinaryReader` and `FileBinaryWriter` are used for that.
So, it's possible to **load and save data files that larger than 2 GB.**
2. QOI textures (de)serialization consumes less memory because of using some shared buffers.
3. `UndertaleData` _(and all `Undertale-` data models)_ is `IDisposable`; **when loading another game data, old data is disposed from memory** _(so less memory is used)_.
4. **_"Writing references in parallel..."_ part of the data saving is a lot faster** - added the chunks type and object index dictionaries, added `IUndertaleSingleChunk` and `IUndertale(Simple)ListChunk` interfaces.
5. Messages of the final part of data saving are describing the process more precisely.
6. `ProfileSaveEvent()` and `UpdateProfile()` are running on different thread, so UI won't freeze on MD5 calculation.

Other changes:

1. `CachedTileDataLoader.Reset()` disposes all cached `Bitmap`s before clearing the dictionaries.
2. `ToString()` overloads of some data models don't throw `NullReferenceException` if `Name` is null.
3. Most of backing fields of the data models are properly named _("\_Name" -> "\_name")_.
4. Code of `UndertaleTags.GetAssetTagID()` is simplified.
5. **Improved (de)serialization methods description.**
7. Fixed glyph size (`Glyph.SourceWidth`/`Glyph.SourceHeight`) description grammar mistake.
8. Added descriptions to some properties.
9. **Fixed some warning/error message boxes not showing up on load or save.**
10. **Embedded textures have read-only width and height properties** _(no UI fields for them yet)_.

### Caveats
Game data loading and saving is now more depending on disk read/write load - if disk will be used while loading/saving the data, then it will process the data longer.

### Notes
Closes #520

### CODE REVIEW

File: UndertaleModLib/Models/UndertaleAnimationCurve.cs [hunk 1]
```diff
@@ -66,11 +68,23 @@ public void Unserialize(UndertaleReader reader, bool includeName)
...
+    }
+
+    /// <inheritdoc/>
+    public void Dispose()
+    {
+        GC.SuppressFinalize(this);
+
+        if (Channels is not null)
+            foreach (Channel channel in Channels)
+                channel?.Dispose();
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleAnimationCurve.cs:81]:
```suggestion
        if (Channels is not null)
        {
            foreach (Channel channel in Channels)
                channel?.Dispose();
         }
```

File: UndertaleModLib/Models/UndertaleEmbeddedTexture.cs [hunk 2]
```diff
@@ -101,25 +103,82 @@ public override string ToString()
...
+            get => _textureBlob;
+            set
+            {
+                _textureBlob = value;
+                OnPropertyChanged();
+            }
+        }
+
+        /// <summary>
+        /// A texture width.
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleEmbeddedTexture.cs:138]:
```suggestion
        /// The width of the texture.
```

File: UndertaleModLib/Models/UndertaleEmbeddedTexture.cs [hunk 3]
```diff
@@ -101,25 +103,82 @@ public override string ToString()
...
+        /// </summary>
+        public int Width => _width;
+
+        /// <summary>
+        /// A texture height.
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleEmbeddedTexture.cs:143]:
```suggestion
        /// The height of the texture.
```

File: UndertaleModLib/Models/UndertaleEmbeddedTexture.cs [hunk 4]
```diff
@@ -101,25 +103,82 @@ public override string ToString()
...
+        /// Frees up <see cref="sharedStream"/> from memory.
+        /// </summary>
+        public static void ClearSharedStream()
+        {
+            sharedStream?.Dispose();
+            sharedStream = null;
+        }
+
+        /// <summary>
+        /// Initializes <see cref="sharedStream"/> with defined initial size.
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleEmbeddedTexture.cs:177]:
```suggestion
        /// Initializes a <see cref="sharedStream"/> with a specified initial size.
```

File: UndertaleModLib/Models/UndertaleEmbeddedTexture.cs [hunk 5]
```diff
@@ -101,25 +103,82 @@ public override string ToString()
...
+        /// </summary>
+        /// <param name="size">Initial size of the stream in bytes</param>
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleEmbeddedTexture.cs:179]: can you make "steam" cref to `sharedStream`?

File: UndertaleModLib/Models/UndertaleEmbeddedTexture.cs [hunk 6]
```diff
@@ -169,16 +231,17 @@ public void Unserialize(UndertaleReader reader)
...
-                    reader.Position = (uint)bufferWrapper.Position;
-                    result.Seek(0, SeekOrigin.Begin);
-                    using Bitmap bmp = QoiConverter.GetImageFromSpan(result.GetBuffer());
-                    using MemoryStream final = new MemoryStream();
-                    bmp.Save(final, ImageFormat.Png);
-                    TextureBlob = final.ToArray();
+                    if (sharedStream.Length != 0)
+                        sharedStream.Seek(0, SeekOrigin.Begin);
+                    BZip2.Decompress(reader.Stream, sharedStream, false);
+                    using Bitmap bmp = QoiConverter.GetImageFromSpan(sharedStream.GetBuffer().AsSpan()[..(int)sharedStream.Position]);
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleEmbeddedTexture.cs:237]: Can you create a local var for `sharedStream.GetBuffer().AsSpan()[..(int)sharedStream.Position]` for readibilty?

File: UndertaleModLib/Models/UndertaleEmbeddedTexture.cs [hunk 7]
```diff
@@ -169,16 +231,17 @@ public void Unserialize(UndertaleReader reader)
...
+                    _width = bmp.Width;
+                    _height = bmp.Height;
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleEmbeddedTexture.cs:239]: Why not use the public Height and Width for this?

File: UndertaleModLib/Models/UndertaleExtension.cs [hunk 8]
```diff
@@ -188,13 +198,27 @@ public override string ToString()
...
     }
+
+    /// <inheritdoc/>
+    public void Dispose()
+    {
+        GC.SuppressFinalize(this);
+
+        if (Functions is not null)
+            foreach (UndertaleExtensionFunction func in Functions)
+                func?.Dispose();
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleExtension.cs:209]:
```suggestion
        if (Functions is not null)
        {
            foreach (UndertaleExtensionFunction func in Functions)
                func?.Dispose();
         }
```

File: UndertaleModLib/Models/UndertaleGameObject.cs [hunk 9]
```diff
@@ -350,18 +356,35 @@ public UndertaleCode EventHandlerFor(EventType type, EventSubtypeDraw subtype, I
...
+    {
+        GC.SuppressFinalize(this);
+
+        _sprite.Dispose();
+        _parentId.Dispose();
+        _textureMaskId.Dispose();
+        PhysicsVertices = new();
+        foreach (var ev in Events)
+            foreach (var subEv in ev)
+                subEv?.Dispose();
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleGameObject.cs:378]:
```suggestion
        foreach (var ev in Events)
        {
            foreach (var subEv in ev)
                subEv?.Dispose();
         }
```

File: UndertaleModLib/Models/UndertaleGeneralInfo.cs [hunk 10]
```diff
@@ -760,10 +782,24 @@ public void Unserialize(UndertaleReader reader)
...
     }
+
+    /// <inheritdoc/>
+    public void Dispose()
+    {
+        GC.SuppressFinalize(this);
+
+        if (Constants is not null)
+            foreach (Constant constant in Constants)
+                constant?.Dispose();
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleGeneralInfo.cs:793]:
```suggestion
        if (Constants is not null)
        {
            foreach (Constant constant in Constants)
                constant?.Dispose();
         }
```

File: UndertaleModLib/Util/QoiConverter.cs [hunk 11]
```diff
@@ -167,31 +188,34 @@ public unsafe static Bitmap GetImageFromSpan(ReadOnlySpan<byte> bytes, out int l
         /// Creates a QOI image as a byte array from a <see cref="Bitmap"/>.
         /// </summary>
         /// <param name="bmp">The <see cref="Bitmap"/> to create the QOI image from.</param>
+        /// <param name="disposeBMP">Whether input <see cref="Bitmap"/> should be disposed.</param>
```
Comment (cgytrus) [inline UndertaleModLib/Util/QoiConverter.cs:191]: qoi _convertion_ doesn't care whether the input should be disposed or not
this should be done by the code that calls the converter, not the converter itself

File: UndertaleModLib/Util/QoiConverter.cs [hunk 12]
```diff
@@ -167,31 +188,34 @@ public unsafe static Bitmap GetImageFromSpan(ReadOnlySpan<byte> bytes, out int l
...
         /// <returns>A QOI Image as a byte array.</returns>
         /// <exception cref="Exception">If there was an error with stride width.</exception>
-        public static byte[] GetArrayFromImage(Bitmap bmp, int padding = 4) => GetSpanFromImage(bmp, padding).ToArray();
+        public static byte[] GetArrayFromImage(Bitmap bmp, bool disposeBMP = true, int padding = 4) => GetSpanFromImage(bmp, disposeBMP, padding).ToArray();
 
         /// <summary>
         /// Creates a QOI image as a <see cref="Span"/> from a <see cref="Bitmap"/>.
         /// </summary>
         /// <param name="bmp">The <see cref="Bitmap"/> to create the QOI image from.</param>
+        /// <param name="disposeBMP">Whether input <see cref="Bitmap"/> should be disposed.</param>
```
Comment (cgytrus) [inline UndertaleModLib/Util/QoiConverter.cs:201]: same thing here

File: UndertaleModLib/Models/UndertaleGeneralInfo.cs [hunk 13]
```diff
@@ -837,6 +884,10 @@ public void Serialize(UndertaleWriter writer)
             }
         }
 
+        /// <summary>
+        /// Deserializes the object from a specified <see cref="UndertaleReader"/>.
+        /// </summary>
+        /// <param name="reader">Where to deserialize from.</param>
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleGeneralInfo.cs:890]: This documentation is incomplete: description is not clear enough what's different betwen `Unserialize(UndertaleReader)` and `uint entryCount` is not documented.

File: UndertaleModLib/Models/UndertaleRoom.cs [hunk 14]
```diff
@@ -162,7 +162,19 @@ public enum RoomEntryFlags : uint
     /// </summary>
     public UndertaleSimpleList<UndertaleResourceById<UndertaleSequence, UndertaleChunkSEQN>> Sequences { get; private set; } = new UndertaleSimpleList<UndertaleResourceById<UndertaleSequence, UndertaleChunkSEQN>>();
 
+    /// <summary>
+    /// Call <see cref="OnPropertyChanged(string)"/> for <see cref="BGColorLayer"/> to update the room background color.<br/>
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleRoom.cs:166]:
```suggestion
    /// Calls <see cref="OnPropertyChanged(string)"/> for <see cref="BGColorLayer"/> in order to update the room background color.<br/>
```

File: UndertaleModLib/Models/UndertaleRoom.cs [hunk 15]
```diff
@@ -162,7 +162,19 @@ public enum RoomEntryFlags : uint
...
+    /// Used for GMS 2 rooms.
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleRoom.cs:167]:
```suggestion
    /// Only used for GameMaker: Studio 2 rooms.
```

File: UndertaleModLib/Models/UndertaleRoom.cs [hunk 16]
```diff
@@ -162,7 +162,19 @@ public enum RoomEntryFlags : uint
...
+    /// </summary>
     public void UpdateBGColorLayer() => OnPropertyChanged("BGColorLayer");
+
+    /// <summary>
+    /// Order the room layers list by depth.
+    /// </summary>
+    /// <param name="selectedLayer">
+    /// Optional argument.<br/>
+    /// Should be present if called from the room editor to re-select the selected layer.
+    /// </param>
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleRoom.cs:177]:
```suggestion
    /// <summary>
    /// Orders the room layers list by depth.
    /// </summary>
    /// <param name="selectedLayer"> Should be present if called from the room editor to re-select the selected layer. </param>
```
Also personally don't quite the selectedLayer's description.

**Review by cgytrus — CHANGES_REQUESTED** [review]

File: UndertaleModLib/Models/UndertaleRoom.cs [hunk 17]
```diff
@@ -352,6 +364,10 @@ public void Unserialize(UndertaleReader reader)
         }
     }
 
+    /// <summary>
+    /// Initialize room by setting <c>ParentRoom</c> property of backgrounds or layers, and calculate room grid size if necessary.
+    /// </summary>
+    /// <param name="calculateGrid">Whether should room grid size be calculated.</param>
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleRoom.cs:370]:
```suggestion
    /// <summary>
    /// Initialize the room by setting <see cref="ParentRoom"/> of Backgrounds or Layers, and optionally calculate the room grid size.
    /// </summary>
    /// <param name="calculateGrid">Whether to calculate the room grid size.</param>
```

Would prefer if Backgrounds and Layers would get a cref as well, its cumbersome from web GH

File: UndertaleModLib/Models/UndertaleRoom.cs [hunk 18]
```diff
@@ -499,7 +547,17 @@ public class Background : UndertaleObject, INotifyPropertyChanged
         /// The y coordinate of the background in the room.
         /// </summary>
         public int Y { get => _y; set { _y = value; OnPropertyChanged(); UpdateStretch(); } }
+
+        /// <summary>
+        /// Whether this background is tiled horizontally.<br/>
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleRoom.cs:552]: Use cref for background pls

File: UndertaleModLib/Models/UndertaleRoom.cs [hunk 19]
```diff
@@ -499,7 +547,17 @@ public class Background : UndertaleObject, INotifyPropertyChanged
...
+        /// <c>0</c> - <see langword="false"/>, <c>1</c> - <see langword="true"/>.
+        /// </summary>
         public int TileX { get; set; } = 1;
+
+        /// <summary>
+        /// Whether this background is tiled vertically.<br/>
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleRoom.cs:558]: Use cref for background pls

File: UndertaleModLib/Models/UndertaleRoom.cs [hunk 20]
```diff
@@ -1496,6 +1630,29 @@ public void Unserialize(UndertaleReader reader)
...
+                        tile?.Dispose();
+                if (Sprites is not null)
+                    foreach (SpriteInstance inst in Sprites)
+                        inst?.Dispose();
+                if (Sequences is not null)
+                    foreach (SequenceInstance inst in Sequences)
+                        inst?.Dispose();
+                if (NineSlices is not null)
+                    foreach (SpriteInstance inst in NineSlices)
+                        inst?.Dispose();
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleRoom.cs:1650]: Please wrap {} after all ifs

File: UndertaleModLib/Models/UndertaleRoom.cs [hunk 21]
```diff
@@ -1521,11 +1678,23 @@ public void Unserialize(UndertaleReader reader)
...
             }
+
+            /// <inheritdoc/>
+            public void Dispose()
+            {
+                GC.SuppressFinalize(this);
+
+                if (Properties is not null)
+                    foreach (EffectProperty prop in Properties)
+                        prop?.Dispose();
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleRoom.cs:1689]: Wrap with {} please

File: UndertaleModLib/Util/QoiConverter.cs (hunk 11 above)
Comment (VladiStep) [reply UndertaleModLib/Util/QoiConverter.cs:191]: I added that for the cases like this:
Line 205 of _"UndertaleEmbeddedTexture.cs"_:
```
writer.Write(QoiConverter.GetSpanFromImage(TextureWorker.GetImageFromByteArray(TextureBlob), true,
                        writer.undertaleData.GM2022_3 ? 0 : 4));
 ```
 To dispose image that was created by `GetImageFromByteArray()`

File: UndertaleModLib/Models/UndertaleTags.cs [hunk 22]
```diff
@@ -7,71 +8,34 @@ namespace UndertaleModLib.Models;
...
+            UndertaleScript => ResourceType.Script,
+            UndertaleFont => ResourceType.Font,
+            UndertaleTimeline => ResourceType.Timeline,
+            UndertaleBackground => ResourceType.Background,
+            UndertaleShader => ResourceType.Shader,
+            UndertaleSequence => ResourceType.Sequence,
+            UndertaleAnimationCurve => ResourceType.AnimCurve,
+            _ => throw new ArgumentException("Invalid resource type!")
+        };
+        IList list = data[resource.GetType()] as IList;
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleTags.cs:34]: Not sure how I feel about this indexer....

**Review by Miepee — COMMENTED** [review]
first pass of rewiew, will need to do several passes with IDE as well, since the web interface isn't that great for this.

File: UndertaleModLib/Util/QoiConverter.cs (hunk 12 above)
Comment (VladiStep) [reply UndertaleModLib/Util/QoiConverter.cs:201]: Same answer as before

File: UndertaleModLib/Models/UndertaleEmbeddedTexture.cs (hunk 2 above)
Comment (VladiStep) [reply UndertaleModLib/Models/UndertaleEmbeddedTexture.cs:138]: Then, isn't _"The texture width"_ better?

File: UndertaleModLib/Models/UndertaleEmbeddedTexture.cs (hunk 7 above)
Comment (VladiStep) [reply UndertaleModLib/Models/UndertaleEmbeddedTexture.cs:239]: For keeping the public properties readonly - update them only with `UpdateSize()` (should be called when `TextureBlob` is changed)

File: UndertaleModLib/Util/QoiConverter.cs (hunk 11 above)
Comment (cgytrus) [reply UndertaleModLib/Util/QoiConverter.cs:191]: then the writer.Write call should be put inside a `using(Bitmap bitmap = ...GetImageFromByteArray...)`
or put only the GetSpanFromImage call in the using and store the span in a variable to make the bitmap get disposed as soon as GetSpanFromImage is done without it having to wait for Write to finish to dispose the bitmap

File: UndertaleModLib/Models/UndertaleTags.cs (hunk 22 above)
Comment (VladiStep) [reply UndertaleModLib/Models/UndertaleTags.cs:34]: What's bad about it exactly?

File: UndertaleModLib/Models/UndertaleEmbeddedTexture.cs (hunk 2 above)
Comment (Miepee) [reply UndertaleModLib/Models/UndertaleEmbeddedTexture.cs:138]: nah that sounds weirder imo

File: UndertaleModLib/Models/UndertaleTags.cs (hunk 22 above)
Comment (Miepee) [reply UndertaleModLib/Models/UndertaleTags.cs:34]: I personally just don't like is that it's not that clear from the outside how an indexer for the data file.

But it exists, and is used in other places, so might as well use it here, i guess.

File: UndertaleModLib/Models/UndertaleRoom.cs (hunk 17 above)
Comment (VladiStep) [reply UndertaleModLib/Models/UndertaleRoom.cs:370]: `UndertaleRoom` doesn't have `ParentRoom` property, so it should stay as `<c>`.

**Pull request closed without merging**